liquidity/threshold_rule.go

```diff
@@ -0,0 +1,229 @@
+	// that we have.
+	currentInbound := balances.incoming * 100 / balances.capacity
+	currentOutbound := balances.outgoing * 100 / balances.capacity
+
+	inboundDeficit := uint64(currentInbound) < uint64(minInbound)
+	outboundDeficit := uint64(currentOutbound) < uint64(minOutbound)
+
+	switch {
+	// If we have too little inbound and too little outbound, we cannot
+	// do anything to help our liquidity situation. This will happen in the
+	// case where we have a lot of pending htlcs on our channels.
+	case inboundDeficit && outboundDeficit:
+		return nil, ReasonNoSurplus
+
+	// If we have too little inbound, but not too little outbound, it is
+	// possible that a loop out will improve our inbound liquidity.
+	case inboundDeficit && !outboundDeficit:
+		out := swap.TypeOut
+		return &out, ReasonImbalanced
+
+	// If we have enough inbound, and too little outbound, it is possible
+	// that we can loop in to improve our outbound liquidity.
+	case !inboundDeficit && outboundDeficit:
+		in := swap.TypeIn
+		return &in, ReasonImbalanced
+
+	// If we have enough inbound and enough outbound, we do not need to
+	// take any actions at present.
+	default:
+		return nil, ReasonLiquidityOk
+	}
+}
+
+// calculateSwapAmount calculates the amount of capacity that we need to shift
+// to improve the balance of our channels, and returns a reason which provides
+// additional context. This function is used in the case where we have surplus
+// liquidity in one direction, and deficit in another. This is the case in which
+// we recommend swaps, if we have deficits on both sides, we cannot swap without
+// further unbalancing, and if we have surplus on both sides, we do not need to.
+//
+// This function calculates the portion of our total capacity we should shift
+// from the surplus side to deficit side without unbalancing the surplus side.
+// This is important, because we do not want to recommend a swap in one
+// direction that just results in our needing to produce a swap in the other
+// direction.
+//
+// If we do have enough surplus, we aim to fall in the midpoint of our two
+// minimums so that we do not tip our channel balance too far in one direction
+// or the other.
+func calculateSwapAmount(capacity, deficitCurrent, deficitMinimum,
+	surplusCurrent, surplusMinimum lnwire.MilliSatoshi) lnwire.MilliSatoshi {
+
+	// Get the minimum shift for our deficit side that we need to swap to
+	// reach our required threshold.
+	required := deficitMinimum - deficitCurrent
+
+	// Get the maximum shift allowed on our surplus side that is possible
+	// before it dips beneath its minimum.
+	available := surplusCurrent - surplusMinimum
+
+	// If the amount of surplus we have is less than the minimum amount we
+	// need to address our deficit, we split the excess between our
+	// directions so that we move towards our desired amount, even if we
+	// can't reach our threshold.
+	if available <= required {
+		return available / 2
```

Comment:
I find this function not so easy to understand, but maybe I think about the problem in a too simplistic way. One question that I have is whether this function is always continuous, no steps?

Isn't the swap amount not just the `midpointDelta`, capped by `available`?